### PR TITLE
feat: parsing & hir of impl blocks

### PR DESCRIPTION
**Pull request by baszalmstra**
This is the first PR to support `impl` blocks for structs. Features included in this PR are:

- Parsing of `impl` blocks. (`ast::Impl` and friends)
- Support `impl` blocks in `ItemTree`.
- Resolve self type of `impl`
- Collect all `impl`s from a package, or for a specific (struct) type.
- Several diagnostics related to `impl` blocks. (See tests in `method_resolution.rs`)
- `Impl` code model to interact with impl blocks

Notable features that are still missing and will be addressed in followup PRs:

- `self` type.
- Code generation.
- Only functions are support in impls. (type alias is missing)
- Language server support
- Docs

### CODE REVIEW

File: crates/mun_hir/src/method_resolution.rs [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        }
+    }
+
+    fn module_id(&self, db: &dyn DefDatabase) -> ModuleId {
+        match self {
+            InherentImplsDiagnostics::LowerDiagnostic(impl_id, _)
+            | InherentImplsDiagnostics::InvalidSelfTy(impl_id)
+            | InherentImplsDiagnostics::IncoherentType(impl_id) => impl_id.module(db),
+            InherentImplsDiagnostics::DuplicateDefinitions(_first, second) => second.module(db),
+        }
```
Comment (Wodann) [inline crates/mun_hir/src/method_resolution.rs:203]: Why do you use the `second` module at this point and not the first?

File: crates/mun_syntax/src/ast/extensions.rs [hunk 2]
```diff
@@ -210,3 +210,24 @@ impl ast::UseTree {
             .any(|it| it.kind() == T![*])
     }
 }
+
+impl ast::TypeAliasDef {
+    pub fn signature_range(&self) -> TextRange {
```
Comment (Wodann) [inline crates/mun_syntax/src/ast/extensions.rs:215]: A comment might help here. What do you mean by "signature" range?

File: crates/mun_syntax/src/ast/generated.rs [hunk 3]
```diff
@@ -100,6 +100,79 @@ impl ArrayType {
...
+        match self.syntax.kind() {
+            FUNCTION_DEF => {
+                AssociatedItemKind::FunctionDef(FunctionDef::cast(self.syntax.clone()).unwrap())
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl AssociatedItem {}
```
Comment (Wodann) [inline crates/mun_syntax/src/ast/generated.rs:146]: Any idea why this empty `impl` block is generated?

File: crates/mun_hir/src/diagnostics.rs [hunk 4]
```diff
@@ -825,3 +824,41 @@ impl Diagnostic for PrivateTypeAlias {
         self
     }
 }
+
+#[derive(Debug)]
+pub struct IncoherentImpl {
```
[the remark this answers is not in the document]
Comment (baszalmstra) [reply crates/mun_hir/src/diagnostics.rs:829]: I renamed them to `ImplForForeignType` to make it clearer.

File: crates/mun_hir/src/ids.rs [hunk 5]
```diff
@@ -90,6 +93,23 @@ pub struct ModuleId {
     pub local_id: LocalModuleId,
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum ItemContainerId {
```
[the remark this answers is not in the document]
Comment (baszalmstra) [reply crates/mun_hir/src/ids.rs:97]: Done

File: crates/mun_hir/src/item_tree/lower.rs [hunk 6]
```diff
@@ -1,27 +1,31 @@
...
 };
 use smallvec::SmallVec;
 use std::{collections::HashMap, convert::TryInto, marker::PhantomData, sync::Arc};
 
 struct ModItems(SmallVec<[ModItem; 1]>);
 
+struct Foo {}
+impl Foo {
+    fn bar() {}
+}
```
[the remark this answers is not in the document]
Comment (baszalmstra) [reply crates/mun_hir/src/item_tree/lower.rs:27]: Yes, removed

File: crates/mun_hir/src/method_resolution.rs (hunk 1 above)
Comment (baszalmstra) [reply crates/mun_hir/src/method_resolution.rs:203]: The error is about a redefinition of a name. So the second definition is to what the error refers.

File: crates/mun_syntax/src/ast/extensions.rs (hunk 2 above)
Comment (baszalmstra) [reply crates/mun_syntax/src/ast/extensions.rs:215]: Added to all implementations (typealias, struct, fn)

File: crates/mun_syntax/src/ast/generated.rs (hunk 3 above)
Comment (baszalmstra) [reply crates/mun_syntax/src/ast/generated.rs:146]: That is due to how we generate code. We only add some functions to this impl if there is stuff to implement. There are more examples of this in the file.

**Review by Wodann — APPROVED** [review]

**Pull request merged**